### PR TITLE
Move most geometry data to C++

### PR DESCRIPTION
**Pull request by smharper**
Here is another PR for the #603 chain.  It introduces the `Cell`, `Universe`, `Lattice`, `RectLattice`, and `HexLattice` types to C++.  It isn't a complete move of those structures to C++, but I think this is about the best trade-off between completeness, number of changes to review, and time.  Plus, this PR is limited to changes which don't break our reproducibility.

Beyond simply translating Fortran to C++, I believe this PR also improves on the structure and cleanliness of the geometry module.  A lot of megalithic subprograms like `read_geomety_xml` and `distance_to_boundary` now send much of the work to type-bound methods.  One place where I was able to take a lot of advantage of C++ features is the new `LatticeIter` type which makes it easy to iterate over lattice universes in a polymorphic fashion.  I've also done a big overhaul of the distribcell logic; it should provide exactly the same results as before, but I think it's much more readable now.

There is one change here that I think is unfortunately less clear than the old Fortran code relating to the region specification.  In Fortran, our region spec contained integer tokens whose absolute value gives the index of the corresponding surface and whose sign gives which side of the surface the region lies on.  However, in C++ one of the surfaces will have an index of 0 which can't be signed.  The way I've gotten around that is by just introducing an off-by-one indexing in the region spec.  Let me know if you see a better option.

Here are a few to-do's that I haven't covered here but I plan to address in later PRs as necessary:
* Find a way to call `Lattice::distance` that works for both `RectLattice` and `HexLattice`.  Currently this is handled by an ugly `select type` in `distance_to_boundary`.  I think the best solution would be to always give this function values from the coordinate level above the lattice tile, but that will likely break our `RectLattice` reproducibility.
* Consider making `LatticeIter` a standards-compliant iterator.  This would mean adding a few operators and defining things like `iterator_category`.  That should allow us to use `LatticeIter` in some standard algorithms but currently we have no need.
* Check the distribcell logic in `fill_offset_tables`.  I think it will fail if there are two different non-nested universes that contain the distribcell.  I believe the new implementation here matches the old one (see `calc_offsets` and how it is called in `prepare_distribcell` for the F90 equivalent).

Finally, there is this mysterious problem that I ran into:  See the `case (FILL_LATTICE)` portion of `write_geometry` in summary.F90.  These two lines of code work fine:
```Fortran
cell_fill = c % fill() + 1
call write_dataset(cell_group, "lattice", lattices(cell_fill)%obj%id())
```
but this one causes a segfault:
```Fortran
call write_dataset(cell_group, "lattice", lattices(c%fill()+1)%obj%id())
```
despite this being fine:
```Fortran
call write_dataset(cell_group, "fill", universes(c%fill()+1)%id)
```
I suspect this is a compiler bug, but it is highly context dependent.  Even just calling `write(*, *) c % fill() + 1` prevents the segfault.  Given it's slippery nature, I doubt we'll be able to reproduce it with a simple example for a GCC bug report.  (I'm running on GCC 7.3.0.)

### CODE REVIEW

File: src/xml_interface.cpp [hunk 1]
```diff
@@ -37,4 +35,28 @@ get_node_value(pugi::xml_node node, const char *name)
   return value;
 }
 
+std::vector<std::string>
+split(const std::string in)
```
Comment (paulromano) [inline src/xml_interface.cpp]: `const` is superfluous since `in` will be copied. Probably what you want is `const std::string& in`

Comment (paulromano) [reply src/xml_interface.cpp]: Also, this function probably should be in some other file since it's not XML-specific.

File: src/lattice.cpp [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+
+  if (check_for_node(lat_node, "outer")) {
+    outer = stoi(get_node_value(lat_node, "outer"));
+  }
+}
+
+//==============================================================================
+
+LatticeIter Lattice::begin()
+{return LatticeIter(*this, 0);}
```
Comment (paulromano) [inline src/lattice.cpp]: Thinking about this a little, I don't think you actually need to define your own `LatticeIter` type. You could just delegate to the iterator of `universes`, i.e.
```C++
{ return this->universes.begin(); }
```
And the same thing for `end`, `rbegin`, and `rend`.

File: src/lattice.cpp [hunk 3]
```diff
@@ -0,0 +1,945 @@
...
+    std::array<double, 2> ll_short {{lower_left[0], lower_left[1]}};
+    write_double(lat_group, "lower_left", ll_short, false);
+    std::array<int, 2> nc_short {{n_cells[0], n_cells[1]}};
+    write_int(lat_group, "dimension", nc_short, false);
+  }
+
+  // Write the universe ids.  The convention here is to switch the ordering on
+  // the y-axis to match the way universes are input in a text file.
+  if (is_3d) {
+    hsize_t nx {static_cast<hsize_t>(n_cells[0])};
```
Comment (paulromano) [inline src/lattice.cpp]: I don't think a static_cast is necessary from `int` to `hsize_t`.

File: src/lattice.cpp [hunk 4]
```diff
@@ -0,0 +1,945 @@
...
+        universes[indx1] = stoi(univ_words[indx2]);
+      }
+    }
+  }
+}
+
+//==============================================================================
+
+int32_t&
+RectLattice::operator[](const int i_xyz[3])
```
Comment (paulromano) [inline src/lattice.cpp]: Will this work for a 2D lattice? (same question goes for hexagonal lattice)

File: src/cell.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+//==============================================================================
+
+extern "C" int32_t n_cells;
+
+class Cell;
+extern std::vector<Cell*> cells_c;
+extern std::unordered_map<int32_t, int32_t> cell_dict;
+
+class Universe;
+extern std::vector<Universe*> universes_c;
```
[the remark this answers is not in the document]
Comment (smharper) [reply src/cell.h]: Sure, thing.  Using `universes` gives us some name collision issues with lattice universes so I've elected for the more specific `global_universes`

File: src/cell.cpp [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+
+  int32_t cell_n_instances(Cell *c) {return c->n_instances;}
+
+  bool cell_simple(Cell *c) {return c->simple;}
+
+  bool cell_contains(Cell *c, double xyz[3], double uvw[3], int32_t on_surface)
+  {return c->contains(xyz, uvw, on_surface);}
+
+  void cell_distance(Cell *c, double xyz[3], double uvw[3], int32_t on_surface,
+                     double &min_dist, int32_t &i_surf)
```
[the remark this answers is not in the document]
Comment (smharper) [reply src/cell.cpp]: Good catch!  I've noticed that this works fine for Fortran interop, but obviously it is outside the standards so we shouldn't do it.

File: src/geometry_aux.cpp [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+      Cell &c = *cells_c[cell_indx];
+      path << "c" << c.id;
+      return path.str();
+    }
+  }
+
+  // The target must be further down the geometry tree and contained in a fill
+  // cell or lattice cell in this universe.  Find which cell contains the
+  // target.
+  std::vector<std::int32_t>::const_reverse_iterator cell_it
```
[the remark this answers is not in the document]
Comment (smharper) [reply src/geometry_aux.cpp]: Doesn't work with GCC 4.8.  See https://stackoverflow.com/questions/2135094/gcc-reverse-iterator-comparison-operators-missing

File: src/lattice.cpp (hunk 2 above)
Comment (smharper) [reply src/lattice.cpp]: It's handy to have `LatticeIter` with the public `indx` attribute for distribcell purposes.  That `indx` value is used for indexing into `Lattice::offsets`

File: src/lattice.cpp (hunk 3 above)
Comment (smharper) [reply src/lattice.cpp]: It's a narrowing conversion so the compiler complains when debug flags are on.

File: src/lattice.cpp (hunk 4 above)
Comment (smharper) [reply src/lattice.cpp]: It will as long as `i_xyz[2] = 0` which will always be returned by `get_indices` for 2D lattices.

**Comment by smharper** [comment]
Regarding the placement of `*` and `&`, I'm not a huge fan of sticking it on the type because it can mislead you when you declare multiple variables at once.  For example, if you define
```C++
PyObject* a, b
```
you might expect both `a` and `b` to be pointers, but only `a` is a pointer.  To make both a pointer, you need
```C++
PyObject *a, *b
```

File: src/lattice.cpp (hunk 2 above)
Comment (paulromano) [reply src/lattice.cpp]: I see, I see

**Comment by paulromano** [comment]
Thanks again for this!

For the placement of `*` or `&`, the easy way around that problem is to just forbid declaring multiple variables on the same line. Anyway, we can leave that discussion for another time.

**Pull request merged**